src/hooks/useDropdownKeyboardNavigation.ts

```diff
@@ -0,0 +1,116 @@
+import React from 'react';
+
+import useIsComponentFocused from './useIsComponentFocused';
+
+export interface KeyboardNavigationProps {
+  container: React.MutableRefObject<HTMLDivElement | null>;
+  disabledIndices?: number[];
+  listLength: number;
+  initialFocusedIndex?: number;
+  onKeyDown?: (event: KeyboardEvent) => void;
+}
+
+interface DropdownKeyboardNavigationState {
+  focusedIndex: number;
+  setFocusedIndex: (index: number) => void;
+  setup: () => void;
+  teardown: () => void;
+}
+
+const useDropdownKeyboardNavigation = ({
+  container,
+  disabledIndices = [],
+  listLength,
+  initialFocusedIndex,
+  onKeyDown,
+}: KeyboardNavigationProps): DropdownKeyboardNavigationState => {
+  const [focusedIndex, setFocusedIndex] = React.useState<number>(-1);
+  const [isInitialNavigation, setIsInitialNavigation] = React.useState(true);
+  const isComponentFocused = useIsComponentFocused(container);
+
+  const getNextIndex = React.useCallback(
+    (direction: 'down' | 'up', index: number) => {
+      switch (direction) {
+        case 'down':
+          return index < listLength - 1 ? index + 1 : 0;
+        case 'up':
+          return index > 0 ? index - 1 : listLength - 1;
+      }
+    },
+    [listLength]
+  );
+
+  const focusOption = React.useCallback(
+    (direction: 'down' | 'up', index: number) => {
+      let nextIndex = getNextIndex(direction, index);
+
+      while (nextIndex !== index && disabledIndices.includes(nextIndex)) {
+        nextIndex = getNextIndex(direction, nextIndex);
+      }
+      setFocusedIndex(nextIndex);
+
+      setIsInitialNavigation(false);
+    },
+    [disabledIndices, getNextIndex]
+  );
+
+  const handleKeyDown = React.useCallback(
+    (event: KeyboardEvent) => {
+      // Handle keyboard events only if current element is focused
+      if (!isComponentFocused()) return;
+
+      onKeyDown && onKeyDown(event);
+      switch (event.key) {
+        case 'ArrowUp':
+          if (isInitialNavigation && typeof initialFocusedIndex === 'number') {
```

Comment:
The `ArrowUp` and `ArrowDown` cases are very similar and could be handled by a single function:
```
const handleArrowKey = (direction: 'up' | 'down', indexModifier: number = 0) => {
    const targetIndex = isInitialNavigation && typeof initialFocusedIndex === 'number'
      ? initialFocusedIndex + indexModifier
      : focusedIndex;

    focusOption(direction, targetIndex);
    event.preventDefault();
};
```
And then `handleArrowKey('up');` or `handleArrowKey('down', -1);`

Reply:
Changed